### DISCUSSION TITLE
M2M using image sequences

### DISCUSSION THREAD

**Discussion by toyxyz** [Mikubill/sd-webui-controlnet#546] [Ideas]
I'm trying to create an animation using multi-controlnet. 
After using the ControlNet M2M script, I found it difficult to match the frames, so I modified the script slightly to allow image sequences to be input and output. I'm very new to Python, so I don't know if I modified it correctly, but it works as intended. Using this script I can img2img a sequence of images from multiple controlnets.

But I can't quite figure out how to change the input image of Img2Img, not controlnet. I want to put a sequence of images in a specific path into the input field of Img2Img one after another. Is there anyone who can give me some advice?

![image](https://user-images.githubusercontent.com/8006000/223854451-ad0323f9-c962-4416-9cab-0d18e0d7e199.png)

Animation created using a modified M2M script.

https://user-images.githubusercontent.com/8006000/223857789-e8d74891-98ea-4446-9a5f-3c97f6d43907.mp4

![image](https://user-images.githubusercontent.com/8006000/223855152-f433012e-c160-46f0-81fd-73c6113ce09a.png)

The ControlNet M2M code I modified is shown below.

```
import copy
import os
import shutil

import cv2
import gradio as gr
import modules.scripts as scripts

from modules import images
from modules.processing import process_images
from modules.shared import opts
from PIL import Image
from pathlib import Path


def get_all_frames_from_path(path):
    if not os.path.isdir(path):
        return None
    frame_list = []
    for filename in sorted(os.listdir(path)):
        if filename.endswith(".jpg") or filename.endswith(".png"):
            img_path = os.path.join(path, filename)
            img = cv2.imread(img_path)
            if img is not None:
                frame_list.append(img)
    return frame_list

def get_min_frame_num(video_list):
    min_frame_num = -1
    for video in video_list:
        if video is None:
            continue
        else:
            frame_num = len(video)
            print(frame_num)
            if min_frame_num < 0:
                min_frame_num = frame_num
            elif frame_num < min_frame_num:
                min_frame_num = frame_num
    return min_frame_num

class Script(scripts.Script):  
    
    def title(self):
        return "controlnet m2m_TEST"

    def show(self, is_img2img):
        return True

    def ui(self, is_img2img):

        ctrls_group = ()
        max_models = opts.data.get("control_net_max_models_num", 1)
        
        input_list = []

        with gr.Group():
            with gr.Accordion("ControlNet-M2M", open = True): 
                with gr.Column():
                    for i in range(max_models):
                        input_path = gr.Textbox(label=f"ControlNet-{i}", placeholder="image sequence path")
                        input_list.append(input_path)
                    tone_image_path = gr.Textbox(label=f"Color_Image", placeholder="image sequence path")
                    output_path = gr.Textbox(label=f"Output_path", placeholder="Output path")

        ctrls_group += tuple(input_list) + (tone_image_path, output_path)

        return ctrls_group


    def run(self, p, *args):

        path = p.outpath_samples
        
        output_path = args[-1]  # get the last argument, which is the output path
        
        if os.path.isdir(output_path):
            shutil.rmtree(output_path)
        os.mkdir(output_path)
        
        tmp_dir = path + "/tmp/" 
        if os.path.isdir(tmp_dir):
            shutil.rmtree(tmp_dir)
        os.mkdir(tmp_dir)
        
        
        video_num = opts.data.get("control_net_max_models_num", 1)

        image_list = [get_all_frames_from_path(image) for image in args[:video_num]]
        
        color_image_list = [get_all_frames_from_path(image) for image in args[-2]]
        

        frame_num = get_min_frame_num(image_list)
        if frame_num > 0:
            output_image_list = []
            for frame in range(frame_num):
                copy_p = copy.copy(p)
                copy_p.control_net_input_image = []
                for video in image_list:
                    if video is None:
                        continue
                    copy_p.control_net_input_image.append(video[frame])
                proc = process_images(copy_p)
                img = proc.images[0]
                images.save_image(img, output_path, f"output_{frame}")
                copy_p.close()


        else:
            proc = process_images(p)
        
        return proc
```

**Comment by LonicaMewinsky** [comment]
This is what I do:
```
copy_p.init_images = [yourPILimage] * p.batch_size
```
Looks awesome so far regardless!

**Reply by toyxyz** [reply]
Thank you! I'll try.

**Comment by toyxyz** [comment]
Thanks to Lonica Mewinsky's advice, I succeeded in automatically putting color image sequences into Img2Img inputs. thank you! I'll have to test it out to see if I can get the results I want.

![image](https://user-images.githubusercontent.com/8006000/224477444-a1e199e6-3c4b-4277-8e93-5b388ba8ca5d.png)

**Comment by lyonnel4** [comment]
Hi. I'd like to use your script version (I prefer to work with sequences of images instead of videos too).
I understand some parts of your script but I don't know enough about Python or Stable Diffusion to understand the advice of Lonica Mewinsky and apply it to the code.
Could you explain what you change please ?

**Reply by toyxyz** [reply]
Here! I'm not familiar with github so I uploaded the script to gumroad. : https://toyxyz.gumroad.com/l/jydvk

**Comment by OedoSoldier** [comment]
You can check my script: https://github.com/OedoSoldier/enhanced-img2img

It can not only input image sequences with ControlNet but also mask sequences, captions from files, etc. And multi-frame rendering is also incluede.

**Reply by toyxyz** [reply]
Thank you! It looks great.

**Comment by toyxyz** [comment]
With the help of many wonderful people, I was able to add the functionality I wanted to the script! 

Below is an animation created by inputting openpose motion and color image sequences separately.

Script : https://toyxyz.gumroad.com/l/jydvk

https://user-images.githubusercontent.com/8006000/226210075-874a8590-a5ca-4913-a998-d9d6b104983c.mp4

Color image / openpose motion
![image](https://user-images.githubusercontent.com/8006000/226210065-13209417-90fe-40e2-a90f-5ffe21bc67b5.png)

**Reply by GeekyGhost** [reply]
I like your script. I've been working on a similar system in the hopes of making a rotoscope animation type of setup. One of the tools I made for it that I use with content creation is this... https://github.com/GeekyGhost/Concactenator It extracts images and audio from video files and stictches them together with or without audio.

**Reply by supwang** [reply]
> With the help of many wonderful people, I was able to add the functionality I wanted to the script!
> 
> Below is an animation created by inputting openpose motion and color image sequences separately.
> 
> Script : https://toyxyz.gumroad.com/l/jydvk
> 
>  s._2.mp4 
> Color image / openpose motion ![image](https://user-images.githubusercontent.com/8006000/226210065-13209417-90fe-40e2-a90f-5ffe21bc67b5.png)

Hi,
I used your ControlNet image sequence 2 image sequence script v2, but why is the image I generated the same as the initial image?
Thank you.
Below is the setting follow your tutorial.
![project](https://github.com/Mikubill/sd-webui-controlnet/assets/4122964/74b1f4c4-aa11-424b-a61d-9ba7caacd56f)

**Reply by toyxyz** [reply]
Denoising strength seems to be too low. Try changing it to a higher value than 0.9.

**Reply by supwang** [reply]
> Denoising strength seems to be too low. Try changing it to a higher value than 0.9.

Thank you. But why does the character's POSE not change?

**Reply by toyxyz** [reply]
A sufficiently high denoising strength is required to change poses.

**Reply by supwang** [reply]
> A sufficiently high denoising strength is required to change poses.

Thank you.
Do you know how to generate png sequence of controlnet openpose detectmap?

**Reply by toyxyz** [reply]
Set Directory for detected maps auto saving in Controlnet settings.

**Reply by supwang** [reply]
> Set Directory for detected maps auto saving in Controlnet settings.

Thank you. But using this method can only generate one piece at a time.
I use OPENPOSE and currently generate images directly from videos in bulk.

Could you add mask functionality to your script?

**Reply by luthis1124** [reply]
I was looking for a way to do this, and eventually gave up and decided I'd get around to writing an extension to do it. I know it is possible because I did it by accident once and have never been able to replicate it.

**Reply by supwang** [reply]
> I was looking for a way to do this, and eventually gave up and decided I'd get around to writing an extension to do it. I know it is possible because I did it by accident once and have never been able to replicate it.

Do you mean generating openpose png sequence detectmap?

**Reply by luthis1124** [reply]
Generating an openpose sequence, yes, actually if you scroll down a bit you'll see where I asked this question.

**Reply by supwang** [reply]
> Generating an openpose sequence, yes, actually if you scroll down a bit you'll see where I asked this question.

I solved the problem.
You can use OpenPose v1.7.0 to generate png sequence.
[https://github.com/CMU-Perceptual-Computing-Lab/openpose/releases/download/v1.7.0/openpose-1.7.0-binaries-win64-gpu-python3.7-flir-3d_recommended.zip](url)
**Command: bin\OpenPoseDemo.exe --video examples\media\test.mp4 --write_images examples\media\images --disable_blending --model_pose COCO --alpha_pose 1**
<img width="1015" alt="2023-05-26 135003" src="https://github.com/Mikubill/sd-webui-controlnet/assets/4122964/19d9a9b3-9b23-43e9-a9aa-7c6735640671">
I saw you asking the same question on Hugging Face, and I already replied to you. :)

**Comment by jAyATJaY** [comment]
Ben racking my brain trying to figure out why the script shows up as a duplicate of the default M2M when I install it. All of the dependencies are met. I can't figure it out.

**Comment by luthis1124** [comment]
I'm really struggling to get the script to work. It seems like the intended use is to do a first run to get the pose images and then use those as input? But when I put any folder in ControlNet-0 I get IndexError: list index out of range. I guess I'm not actually understanding how to use this step by step.

I must have read through your gumroad page 20 times trying to figure out what I'm doing wrong. Any advice?

**Reply by toyxyz** [reply]
> I'm really struggling to get the script to work. It seems like the intended use is to do a first run to get the pose images and then use those as input? But when I put any folder in ControlNet-0 I get IndexError: list index out of range. I guess I'm not actually understanding how to use this step by step.
> 
> I must have read through your gumroad page 20 times trying to figure out what I'm doing wrong. Any advice?

First you need to prepare the image sequence for controlnet. The name must be numbered in order, such as a-000, a-001. When using a color image sequence, prepare the same number as the controlnet image. 
![image](https://user-images.githubusercontent.com/8006000/234269496-1ea220eb-dc61-4328-9d45-12070a9c8b3e.png)
And feed the first color image to the img2img input.
![image](https://user-images.githubusercontent.com/8006000/234269901-ebe6326d-1fb8-4285-a48c-a091c49c4474.png)
Now enter the path of the image sequences you have prepared.
![image](https://user-images.githubusercontent.com/8006000/234269983-cde84e49-3e1d-43ee-b36f-29f7cb2e8966.png)
The result is:
![image](https://user-images.githubusercontent.com/8006000/234270274-970f1de2-6965-4d26-a691-0a907d05e778.png)

**Reply by luthis1124** [reply]
Great thank you! I managed to get it working. Kind of. I was making a few mistakes, most importantly not converting .bmp to .png which is why I was getting list index error. Also I was using batch, not just single img2img.
So please correct me if I'm wrong, this script takes as input:
one single initial starting image ie the first image that was used to get poses (OR is this only the colour image?)
one folder for pose images
optionally a folder for colour images

So I just have 2 more questions:
How to generate all the pose images from pngs? I can do it one at a time but there must be a way to do them as a batch, because I definitely have done it before by accident.

So this script will ONLY use a _single image_ (the initial image) as a base to generate all the next images? Because it seemed like it's the same as batch, like it would use each frame + colour + pose + previous frame to generate. But as written here it is just colour + pose + previous frame.

**Reply by luthis1124** [reply]
Actually, reading through the script it looks like you're using the first image to find the rest of the set so, I just need to know how to generate the openpose images now.

**Reply by toyxyz** [reply]
There are two ways to prepare this image for the open pose. Using the preprocessor provided by the extension, and using a 3D model with a similar appearance, as in my case. Same goes for my script. If you entered the normal human image sequence into Controlnet 0, set the preprocessor to open pose, and if you prepared an open pose image, set it to none.

This script uses a different color image for every frame if the user inputs the path of the color image sequence.(A) Otherwise, the single image entered in img2img is used for the whole. (B)

An option to mix methods A and B is also provided. (Blend color image)

**Reply by toyxyz** [reply]
> Actually, reading through the script it looks like you're using the first image to find the rest of the set so, I just need to know how to generate the openpose images now.

If you input a normal human image sequence and set the preprocessor to open pose, you get the following result.
![image](https://user-images.githubusercontent.com/8006000/234740815-e3d4e807-1def-45a2-8ca4-4fa03e25281b.png)

**Comment by luthis1124** [comment]
Thank you for continuing to answer by questions!
I just need to know how to generate these:
![image](https://user-images.githubusercontent.com/55561775/234752455-0bc69b49-134c-418c-a88c-3e0683cc0c13.png)
To use here:
![image](https://user-images.githubusercontent.com/55561775/234752591-e02c2aee-5a0f-4817-b8f4-c49cb687abbc.png)
From your example steps above.

**Reply by toyxyz** [reply]
Those pose images were rendered using this 3D model for Blender. https://toyxyz.gumroad.com/l/ciojz
